### ISSUE TITLE
Issue with feature functions in chapter8/optimal_exercise_bi.py

### ISSUE THREAD

**Issue by guilhermebrunetti** [TikhonJelvis/RL-book#178]
Dear professors,

I believe there is an issue with the feature functions used in the chapter8/optimal_exercise_bi.py, in the lines 200-202:

```
ffs += [(lambda s: np.log(1 + np.exp(-s.state / (2 * strike))) *
            lagval(s.state / strike, ident[i]))
            for i in range(num_laguerre)]
```

It should create different functions, each one being a different Laguerre polynomial multiplied by a softmax function.

However, it seems that it is creating the same function multiple times.

Searching for this issue on the internet, I found a nice post in Stackoverflow (https://stackoverflow.com/questions/6076270/lambda-function-in-list-comprehensions), where it explains the problem (a sort of edge case in Python, where the loop variable, the ``i`` in the code above, is captured by reference instead of by value).

It seems one possible solution would be to explicitly capture the loop variable inside the function as an extra argument, like below:

```
ffs += [(lambda s, coeff=i: np.log(1 + np.exp(-s.state / (2 * strike))) *
            lagval(s.state / strike, ident[coeff]))
            for i in range(num_laguerre)]
```

It might be that this same issue happens in other places in the codebase. The solution would be similar.

Would you please let me know what you think about this?

Thanks for the code examples and the book.

**Comment by TikhonJelvis** [comment]
Thanks for catching that! This specific example is fixed (in #179); we'll do an editing pass through all the code examples at some point, which should help catch the rest. In the meantime, feedback like this is really helpful.

I recall that, in older versions of Python, `i` would have been in scope *outside the list comprehension* :exploding_head:. Looks like at least *that* has been fixed, probably in the transition from 2 to 3.

An alternative would be to have a named function:

``` python
def laguerre(i):
    return lambda s: (np.log(1 + np.exp(-s.state / 2 * strike)) * \
            lagval(s.state / strike), ident[i])

ffs += [laguerre(i) for i in range(num_laguerre)]
```

(I haven't tested this exact code, so there might be some typos.)

Which version seems clearer? I don't have a feel for which approach would be more pythonic.

**Comment by coverdrive** [comment]
Fixed
